### PR TITLE
Add in SatFlow losses

### LINKED ISSUE #13: Add loss functions from SatFlow
Move loss functions from SatFlow over here, so they can be shared

### PR DESCRIPTION
**Pull request by jacobbieker**
# Pull Request Template

## Description

Adds in all the losses from SatFlow, and the `get_loss` convience function

Fixes #13  (issue)

## How Has This Been Tested?

Unit tests

- [ ] No
- [x] Yes

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: nowcasting_utils/models/loss.py [hunk 1]
```diff
@@ -43,3 +51,310 @@ def get_mse_exp(self, output, target):
...
+        """
+        super(SSIMLoss, self).__init__()
+        self.convert_range = convert_range
+        self.ssim_module = SSIM(**kwargs)
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor):
+        if self.convert_range:
+            x = torch.div(torch.add(x, 1), 2)
+            y = torch.div(torch.add(y, 1), 2)
+        return 1.0 - self.ssim_module(x, y)
```
Comment (peterdudfield) [inline nowcasting_utils/models/loss.py:72]: why is it 1 - here?

File: nowcasting_utils/models/loss.py [hunk 2]
```diff
@@ -43,3 +51,310 @@ def get_mse_exp(self, output, target):
...
+        Args:
+            convert_range: Whether to convert from -1,1 to 0,1 as required for SSIM
+            **kwargs: Kwargs for the ssim_module
+        """
+        super(SSIMLossDynamic, self).__init__()
+        self.convert_range = convert_range
+        self.ssim_module = MS_SSIM(**kwargs)
+
+    def forward(self, curr_image: torch.Tensor, x: torch.Tensor, y: torch.Tensor):
+        if self.convert_range:
```
Comment (peterdudfield) [inline nowcasting_utils/models/loss.py:111]: worth adding function descirption, for example what curr_image is

maybe worth using 'current_image' - fully verbose variables

File: nowcasting_utils/models/loss.py (hunk 1 above)
Comment (jacobbieker) [reply nowcasting_utils/models/loss.py:72]: Because SSIM has a score of 1 if its a perfect match, so this makes it minimize the value, with a loss of 0 if the target and output are identical

File: nowcasting_utils/models/loss.py [hunk 3]
```diff
@@ -43,3 +51,310 @@ def get_mse_exp(self, output, target):
...
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, x, real_flag):
+        if real_flag is True:
+            x = -x
+        return F.relu(1.0 + x).mean()
+
+
+class FocalLoss(nn.Module):
```
Comment (peterdudfield) [inline nowcasting_utils/models/loss.py:221]: worth splitting into seperate file? 
Might be worth putting the TV Loss also in a seperate file, just to keep it tidy

File: nowcasting_utils/models/loss.py [hunk 4]
```diff
@@ -43,3 +51,310 @@ def get_mse_exp(self, output, target):
...
+
+        if self.size_average:
+            loss = loss.mean()
+        else:
+            loss = loss.sum()
+        return loss
+
+
+def get_loss(loss: str = "mse", **kwargs) -> torch.nn.Module:
+    if isinstance(loss, torch.nn.Module):
```
Comment (peterdudfield) [inline nowcasting_utils/models/loss.py:319]: nice i like this function, 

Could you add function description  please

File: tests/models/test_loss.py [hunk 5]
```diff
@@ -59,3 +59,72 @@ def test_mse_exp_rand():
...
+    "tv",
+    "total_variation",
+    "ssim_dynamic",
+    "gdl",
+    "gradient_difference_loss",
+    "weighted_mae",
+    "weighted_mse"
+])
+def test_get_loss(loss_name):
+    loss = get_loss(loss_name)
```
Comment (peterdudfield) [inline tests/models/test_loss.py:82]: nice

File: tests/models/test_loss.py [hunk 6]
```diff
@@ -59,3 +59,72 @@ def test_mse_exp_rand():
...
+def test_tv_loss():
+    loss = get_loss("tv")
+    output = torch.randn((2, 12, 512, 512))
+    out = loss(output)
+    assert out > 0
+
+
+def test_missing_loss():
+    with pytest.raises(AssertionError):
+        loss = get_loss("made_up_metric")
```
Comment (peterdudfield) [inline tests/models/test_loss.py:107]: nice

**Review by peterdudfield — APPROVED** [review]
Few little changes, but looks good otherwise

File: nowcasting_utils/models/loss.py (hunk 1 above)
Comment (peterdudfield) [reply nowcasting_utils/models/loss.py:72]: ah nice, is it ok to add that in as a comment, then someone else reading it wont get confused

File: nowcasting_utils/models/loss.py [hunk 7]
```diff
@@ -20,20 +26,24 @@ def __init__(self, decay_rate: Optional[int] = None, forecast_length: int = 6):
...
+            [math.exp(-self.decay_rate * i) for i in range(0, self.forecast_length)]
+        )
 
         # normalized the weights
         self.weights = weights / weights.sum()
 
         # move weights to gpu is needed
-        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
         self.weights = self.weights.to(device)
```
Comment (JackKelly) [inline nowcasting_utils/models/loss.py:47]: Does manually moving the weights to the GPU play nicely with Pytorch Lightning?  The [PyTorch Lightning docs](https://pytorch-lightning.readthedocs.io/en/stable/advanced/multi_gpu.html) say not to use `.to(device)` :)  (although I'm not sure what the alternative is here!)

File: nowcasting_utils/models/losses/FocalLoss.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        self,
+        apply_nonlin=None,
+        alpha=None,
+        gamma=2,
+        balance_index=0,
+        smooth=1e-5,
+        size_average=True,
+    ):
+        """
+        Copied from: https://github.com/Hsuxu/Loss_ToolBox-PyTorch/blob/master/FocalLoss/FocalLoss.py
```
Comment (JackKelly) [inline nowcasting_utils/models/losses/FocalLoss.py:17]: github.com/Hsuxu/Loss_ToolBox-PyTorch doesn't have an explicit license... Should we ask hsuxu if we can copy this code?  And, if so, how hsuxu would like to be attributed?

File: nowcasting_utils/models/losses/FocalLoss.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        elif isinstance(alpha, float):
+            alpha = torch.ones(num_class, 1)
+            alpha = alpha * (1 - self.alpha)
+            alpha[self.balance_index] = self.alpha
+
+        else:
+            raise TypeError("Not support alpha type")
+
+        if alpha.device != logit.device:
+            alpha = alpha.to(logit.device)
```
Comment (JackKelly) [inline nowcasting_utils/models/losses/FocalLoss.py:74]: Whilst the Lightning docs discourage `to(device)`, this code is probably OK because it feels functionally equivalents to PL's recommendation of `alpha = alpha.type_as(logit)`

File: nowcasting_utils/models/losses/StructuralSimilarity.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
+import torch
+from pytorch_msssim import SSIM, MS_SSIM
+from torch import nn as nn
+
+
+class SSIMLoss(nn.Module):
```
Comment (JackKelly) [inline nowcasting_utils/models/losses/StructuralSimilarity.py:6]: Please add a brief docstring (even if it just says that SSIM stands for Structural Similarity Index Measure, maybe with a link to https://en.wikipedia.org/wiki/Structural_similarity :)

File: nowcasting_utils/models/losses/TotalVariationLoss.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
+import torch
+from torch import nn as nn
+
+
+def tv_loss(img, tv_weight):
+    """
+    Taken from https://github.com/chongyangma/cs231n/blob/master/assignments/assignment3/style_transfer_pytorch.py
```
Comment (JackKelly) [inline nowcasting_utils/models/losses/TotalVariationLoss.py:7]: https://github.com/chongyangma/cs231n lacks an explicit license.  Please ask chongyangma if they're happy for us to copy their code, and how they'd like to be attributed :)

(Sorry for being a pedant about licenses!)

**Review by JackKelly — COMMENTED** [review]
Looks great!  It's so awesome to see all these losses in one place, and so nicely implemented - great work!

File: nowcasting_utils/models/loss.py (hunk 7 above)
Comment (jacobbieker) [reply nowcasting_utils/models/loss.py:47]: It does work okay, I noticed it fails when running the tests locally, where there is an available GPU, but the tests are running on the GPU. @peterdudfield and I chatted about it a bit, the other option seems to add a call to `.type_as(x)` in the computation, which would then run everytime its called, which might be slower, although we probably just need to profile it to check.

File: nowcasting_utils/models/losses/FocalLoss.py (hunk 8 above)
Comment (jacobbieker) [reply nowcasting_utils/models/losses/FocalLoss.py:17]: I've switched it out to an explicitly MIT licensed version!

File: nowcasting_utils/models/losses/TotalVariationLoss.py (hunk 11 above)
Comment (jacobbieker) [reply nowcasting_utils/models/losses/TotalVariationLoss.py:7]: I have found an Apache Licensed version that also has some other nice features, so swapping it out for that!

File: nowcasting_utils/models/loss.py (hunk 7 above)
Comment (JackKelly) [reply nowcasting_utils/models/loss.py:47]: Yeah, it might be worth trying `.type_as(x)` to see if it slows things down.  I'd hope the PyTorch folks have implemented `.type_as(x)` to be as quick as possible :)
